File: js/creational/singleton.js
class User {
    constructor(name, age) {
        if (User.instance) return User.instance

        this.name = name
        this.age = age

        User.instance = this
    }
}

const user_1 = new User('Oleksandr', 24)
const user_2 = new User('Sasha', 22)

console.log(user_1)
console.log(user_2)

console.log(user_1 === user_2)

File: python/creational/singleton.py
class User:
    __instance = None

    def __new__(cls, *args, **kwargs):
        if not isinstance(cls.__instance, cls):
            cls.__instance = super().__new__(
                cls)  # This is a call to the parent's __new__ method, which creates a new object of class cls (e.g. User) and returns it.

        return cls.__instance

    def __init__(self, name, age):
        self.name = name
        self.age = age


user1 = User('Oleksandr', 24)
user2 = User('Sasha', 22)

print(user1.name)
print(user2.name)
